### ISSUE TITLE
[Feature Request] Dynamic DNN creation

### ISSUE THREAD

**Issue by jpdoyle** [davisking/dlib#149]
Currently the DNN interface is very nicely modularized (I like the add_layer<> template, it's clever), but it currently seems impossible to dynamically build a DNN at run time. It looks like you actually aren't too far from this being possible, though. If some runtime-polymorphic version of SUBNET could be created, with a version of add_layer<> that inherits from that SUBNET class, most of the current code could probably support a built-at-run-time network with minimal effort.

Is there a specific reason you chose to restrict network construction to compile-time? It seems to limit your ability to use existing networks that are in formats like .caffemodel

**Comment by e-fominov** [comment]
This question(my) was answered [here](https://sourceforge.net/p/dclib/discussion/442517/thread/d95a42f4/)

I am using this Dlib DNN for a while, and now I am completely sure that I do not need polimorphic nets. Train your brains to think in C++ and you will improve your results much :)

**Comment by jpdoyle** [comment]
> Train your brains to think in C++ and you will improve your results much :)

This is not a helpful way to sell this particular design choice. C++ has 2 main types of polymorphism (templates being one of them), and claiming that using one or the other exclusively is "thinking in C++" doesn't seem like a good engineering approach.

> The target audience is professional
software engineers who want to incorporate DNN technology into C++
applications rather than grad students who want to run experiments.

This line from the thread you linked is rather suspect to me. A professional software engineer ought to take advantage of the freely available, pre-trained networks available through things like Caffe's Model Zoo. Importing networks from other frameworks _would not require code generation_ if even a bare-bones run-time-polymorphic version were implemented. I think it is much more pragmatic to support template-based networks for hardened, high-performance network architectures, and dynamic networks for training and testing.

**Comment by davisking** [comment]
It's not so straight forward to support runtime polymorphic networks given
the existing code base.  Obviously this makes importing caffe models
problematic but there are a bunch of other considerations.  For instance,
importing caffe models at some level foists the complexities and design
problems of caffe onto software that supports reading them.

For example, if you write a new layer type, how does that layer type get
automatically serialized when saving a network?  The template system in use
now makes that automatic.  But with runtime polymorphism you need some kind
of global registry/factory thing that exists somewhere in dlib and performs
mappings between magic strings and classes.  This means that users who want
to implement new layer types need to edit the dlib code.  A major design
goal of this API is to let users create new loss layers, computational
layers, and solvers without needing to understand or even look at the dlib
internals.  A lot of the API decisions are based on what makes the
interface a user needs to implement for new layer creation simple, and if
you look at it you will find that it's far simpler than the interface in
caffe, while also being very general.

Also, the entire structure of template alias statements and tag
introspection doesn't work with runtime polymorphism.

Since it takes days to train these models, needing to recompile to change
the network architecture isn't a big deal.  You recompile every few days.
The main downside is not being able to import models from other tools.  But
this is a short term problem which is easily remedied by putting more dlib
models on the internet.  Supporting import from other tools like caffe is a
never ending odesy of varying and incompatible formats and design
decisions.

What models from caffe do you want to use?

**Comment by e-fominov** [comment]
@davisking, DNN module is released and will be interesting to many developers. I think this question (dynamic DNN) should be added to FAQ. Also explanation of missing MSVC support of DNN should be added too

Btw, is there any way to make PR into FAQ?

**Comment by davisking** [comment]
@e-fominov, yeah you are right.  I keep repeating these explanations about serialization and whatnot.  

Also, I would like to get a list of popular caffe models so I can train them.  So far some people say "I want to load caffe models" but I'm not sure if they are just saying that to be contrary on the internet or there really are interesting usecases for some of the pretrained models.   For perspective, all the vision researchers I know are going to be training their own models, so this kind of thing is a non-issue.  I don't understand how a serious computer vision person can say they aren't training their own models. :/

**Comment by davisking** [comment]
Oh and the FAQ is in this XML file https://github.com/davisking/dlib/blob/master/docs/docs/faq.xml
